README.md
# Bot Binance con RSI

* Este bot está diseñado para realizar trades en la plataforma Binance.
* Utiliza Python 3.5 o superior
* Actualmente utiliza el [indicador RSI](https://academy.binance.com/es/articles/what-is-the-rsi-indicator) para calcular si realiza o no un trade.
* No es de mi total autoría, está basado en otro desarrollo.. (En cuanto recuerde la documentación/video lo pondré)

## Disclaimer!

* Este bot fue hecho con propósitos de aprendizaje y si bien puede utilizarse para realizar trade, se recomienda hacerlo con cautela por los siguientes motivos:
1. No ha sido probado de manera extensa.
2. El indicador RSI por si solo, puede generar compras/ventas en puntos previos al máximo de cada período, por lo cual es ideal poder ocuparlo con otros indicadores

## Installation

Utilizar el administrador de paquetes PIP [pip](https://pip.pypa.io/en/stable/) para instalar los paquetes necesarios, que se encuentran en el archivo requirements.txt utilizando el siguiente comando

```powershell
    # Directament con PIP
    pip install -r requirements.txt

    # En caso de ambiente virtual o PIP no es variable de entorno invocar mediante python de alguna de las siguientes opciones:
    python -m pip install -r requirements.txt

    #O también
    py -m pip install -r requirements.txt
```

## Configuración de uso
* Esta versión 0.1 tiene pre-configurados los parámetros por defecto que utiliza el algoritmo RSI. En caso de requerir customización, se debe hacer en el código en los siguientes parámetros:

```python    
    RSI_PERIOD = 14
    RSI_OVERBOUGHT = 70
    RSI_OVERSOLD = 30
```

* Dependiendo del **símbolo** y del **tiempo** que se desee recolectar los datos, se debe modificar la siguiente línea con el parámetro indicado en la documentación:
1. Para el tiempo: Referir a la documentación de [Binance API](https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#klinecandlestick-streams)
2. Para el símbolo de trade: Tomar de Binance según moneda en análisis. Ej: BNB/USDT se remueve el "/" y queda BNBUSDT 

```python
    SOCKET = "wss://stream.binance.com:9443/ws/SIMBOLO_DE_TRADE@kline__PONER_ACÁ_TIEMPO"

    #ejemplo
    SOCKET = "wss://stream.binance.com:9443/ws/BNBUSDT@kline_1m"
```

* Para el trade además deben ser configurados los siguientes parámetros:

```python
    TRADE_SYMBOL = 'ACÁ_VA_SÍMBOLO_DE_LA_MONEDA'
    TRADE_QUANTITY = 'ACÁ_VA_EL_MONTO_A_TRADEAR'
```

* En el archivo config.py se deben ingresar la ApiKey y ApiSecret, los cuales se deben obtener del sitio web de Binance siguiendo la documentación [¿Cómo crear una clave API?](https://www.binance.com/es/support/faq/360002502072)

```python
    API_KEY = "Entre estas comillas va tu API KEY"
    API_SECRET = "Entre estas comillas va tu API SECRET"
```

## Referencias
* [Binance Api]()
* [Indicador RSI](https://academy.binance.com/es/articles/what-is-the-rsi-indicator)
* [WebSocket Streams](https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md)

CHANGELOG.md
# CHANGELOG

A continuación se indican los cambios realizados en el Bot RSI Binance

## Unreleased

- Se añadirán registros json de RSI calculados y de las órdenes emitidas, para mejor lectura. La idea es dejar registros y limpiar los outputs en consola.

## Versión 0.2

### Added
- Se incluye archivo **CHANGELOG.md**
- Colores para destacar eventos importantes (Conexión iniciada, finalizada, compra y venta)
- Sonidos para notificar eventos importantes (Conexión iniciada, compra y venta)
- Se genera registro json de los puntos de cierre (closes) en directorio: data/closes/closes.json. Esto con motivo de facilitar lectura de eventos en consola.

### Changed

-  Se corrige error en envío de órdenes de compra.
- 

## Versión 0.1

### Added

- Archivo **README.md** incluye forma de uso e instalación.
- Se incluye archivo **requirements.txt**, con los paquetes que utiliza el bot.

bot.py
import websocket, json, pprint, talib, numpy
import config
from binance.client import Client
from binance.enums import *
import os

#Evita que pygame despliegue bienvenida en consola
from os import environ
environ['PYGAME_HIDE_SUPPORT_PROMPT'] = '1'
import pygame

os.system("color")

pygame.init()

#set de audios en variables
b = pygame.mixer.Sound("sounds/buy.mp3")
s = pygame.mixer.Sound("sounds/sell.mp3")

SOCKET = "wss://stream.binance.com:9443/ws/thetabnb@kline_1m"

### Parámetros para cálculo RSI y órdenes Compra/Venta
RSI_PERIOD = 14
RSI_OVERBOUGHT = 70
RSI_OVERSOLD = 30
TRADE_SYMBOL = 'THETABNB'
TRADE_QUANTITY = 0.05

### Array de cierres closes
closes = []

### En posición de compra
in_posicion = True

### Client
client = Client(config.API_KEY, config.API_SECRET)

### Función para enviar orden (compra/venta)
def order(side, quantity, symbol, order_type=ORDER_TYPE_MARKET):
    try:
        print("Enviando Orden")
        order = client.create_order(symbol=symbol, side=side, type=order_type, quantity=quantity)
        print(order)
    except Exception as e:
        print("Ha ocurrido un error al enviar la orden - {}".format(e))
        return False

    return True


def on_open(ws):
    print('\x1b[0;31;43m' + 'System: [Openned Conecction]' + '\x1b[0m' '\n')
    pygame.mixer.Sound.play(b)
    pygame.mixer.music.stop()

def on_close(ws):
    print('\n')
    print('\x1b[0;31;43m' + 'System: [Closed Conecction]' + '\x1b[0m')

def on_message(ws, message):
    global closes, in_posicion, b, s

    #print("received message")
    json_message = json.loads(message)
    #pprint.pprint(json_message)

    candle = json_message['k']
    
    is_candle_closed = candle['x']
    close = candle['c']

    if is_candle_closed:
        print("Candle close at {}".format(close))
        closes.append(float(close))
        json_closes = json.dumps(closes)
        closesFile = open("data/closes/closes.json", "w")
        closesFile.write(json_closes)
        closesFile.close()
    
        ### Cálculo de RSI
        if len(closes) > RSI_PERIOD:
            np_closes = numpy.array(closes)
            rsi = talib.RSI(np_closes, RSI_PERIOD)
            print("RSI Calculados")
            print(rsi)
            last_rsi = rsi[-1]
            print("El RSI actual es {}".format(last_rsi))

            if last_rsi > RSI_OVERBOUGHT:
                if in_posicion:
                    print('\x1b[5;37;42m' + "#### VENDE ####" + '\x1b[0m')
                    # Acá va lógica de venta
                    order_succeeded = order(SIDE_SELL, TRADE_QUANTITY, TRADE_SYMBOL)
                    pygame.mixer.Sound.play(s)
                    pygame.mixer.music.stop()
                    if order_succeeded:
                        in_posicion = False
                else:
                    print("Está Overbought pero ya vendiste.")
            
            if last_rsi < RSI_OVERSOLD:
                if in_posicion:
                    print("Está Oversold, pero ya compraste.")
                else:
                    print('\x1b[5;37;42m' + "#### COMPRA ####" + '\x1b[0m')
                    # Acá va lógica de compra
                    order_succeeded = order(SIDE_BUY, TRADE_QUANTITY, TRADE_SYMBOL)
                    pygame.mixer.Sound.play(b)
                    pygame.mixer.music.stop()
                    if order_succeeded:
                        in_posicion = True

ws = websocket.WebSocketApp(SOCKET, on_open=on_open, on_close=on_close, on_message=on_message)
ws.run_forever()